### PR TITLE
nixos/gitlab: Add support for secure secrets and more

### PR DESCRIPTION
**Pull request by talyz**
###### Motivation for this change
Improve the gitlab module in various ways:
- Add support for storing secrets in files (even arbitrary secrets in extraConfig)
- Split preStart into a privileged and unprivileged part (fixing issue #53852 for the gitlab module)
- Make initialization more robust
- Fix missing ca_file for SMTP (upstream issue 790)

This does, as far as I can tell, what the unmaintained #31358 set out to do, and more. I takes a slightly different approach to the problem by mostly using jq to put secrets in place.

To split preStart I implemented support for this in the systemd module rather than hack it into the gitlab module so that other modules can benefit from it. This should solve issues where PermissionsStartOnly cannot be replaced with only tmpfiles.

###### Needs testing
- Upgrading from an existing gitlab setup on 19.03 to this

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [x] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nix-review --run "nix-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [x] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

###### Notify maintainers

cc @globin @etu @adisbladis @Infinisil @lheckemann @nh2 @aanderse

### CODE REVIEW

**Comment by adamtulinius** [comment]
All the work related to systemd should probably be a separate pull request.

**Comment by talyz** [comment]
Hm, I was advised that it probably shouldn't, since the gitlab work depends on it, but maybe that's incorrect? I can of course split it into two separate PRs, but the gitlab PR would be broken until the systemd PR is merged, but maybe that's okay?

**Comment by adamtulinius** [comment]
> Hm, I was advised that it probably shouldn't, since the gitlab work depends on it, but maybe that's incorrect?

I just figured it would be easier to review as two separate things, since it's probably not the same people responsible for reviewing the gitlab and systemd parts. Hopefully somebody more knowledgeable can chip in.

**Comment by talyz** [comment]
Yeah, I'm good with either, so if anyone feels strongly that it should be split, I'll split it.

File: nixos/modules/services/misc/gitlab.nix
Comment (aanderse) [inline nixos/modules/services/misc/gitlab.nix]: It seems to me that almost all of the code in here can (and should) be taken care of in a declarative way.

- Any password files should already have read permissions for the user that requires them.
- Databases should be provisioned with `services.postgresql.ensureDatabases` and database users with `services.postgresql.ensureUsers`.
- I'm not familiar with `pgsql` extensions, but it seems like maybe there should be a `services.postgresql.ensureExtensions` option...?

I'll also mention that I'm a fan of a `services.foobar.database.createLocally` option to indicate that the sysadmin wants a database provisioned for them.

File: nixos/modules/system/boot/systemd-unit-options.nix
Comment (aanderse) [inline nixos/modules/system/boot/systemd-unit-options.nix]: I fear this would encourage contributors to write more imperative code, instead of the declarative way that should be preferred in modules. I acknowledge there are cases where a script requires full privileges to run, but suggest that the situations where this is the case are few enough that the cons of this change outweigh the pros.

This comment applies to all `*FullPrivileges` options introduced.

File: nixos/modules/services/misc/gitlab.nix
Comment (aanderse) [inline nixos/modules/services/misc/gitlab.nix]: I think `password` options don't provide much value, even for testing. Were a `passwordFile` option to document how easy it is to write something like `passwordFile = pkgs.writeText "insecure-password.txt" "just-for-testing";` I think that should be sufficient.

That being said, I have no idea how popular/unpopular my idea is :man_shrugging:

File: nixos/modules/services/misc/gitlab.nix
Comment (aanderse) [inline nixos/modules/services/misc/gitlab.nix]: With the `gitea` module secrets are automatically generated and stored in the `stateDir`, which a sysadmin is then free to change. Is there value in allowing the sysadmin to generate this key upfront? I would suggest this option should be removed entirely, and in the edge case that a sysadmin actually needs to specify the key they can go into `stateDir` and do so.

File: nixos/modules/services/misc/gitlab.nix
Comment (aanderse) [inline nixos/modules/services/misc/gitlab.nix]: If there were a `services.gitlab.database.createLocally` option you could make `"postgresql.service"` optional here as otherwise you wouldn't have a dependency on a service which is likely not even running.

**Review by aanderse — CHANGES_REQUESTED** [review]
First - thank you for tackling this! The module could really use some love. I'm really glad you have taken this on as the end goal is fantastic. I have some opinions on how you're trying to achieve that goal. I hope we can have some productive discussion on the topic.

File: nixos/modules/services/misc/gitlab.nix
Comment (talyz) [reply nixos/modules/services/misc/gitlab.nix]: While I do agree that it would be preferable to do everything in a declarative way, I'm not entirely sure this is possible. I'll just reply to the points in the same order:
- I'm not sure what this is in reference to?
- The problem is that there's no way, afaict, to create a user with a password in a secure fashion using the existing postgres service options. Maybe the postgres module should provide a way, though?
- Same here, as you say, it's not currently possible, but maybe it could be implemented?

Yeah, I guess we could introduce a `createLocally` option and rely on it instead of checking `databaseHost`.

File: nixos/modules/system/boot/systemd-unit-options.nix
Comment (talyz) [reply nixos/modules/system/boot/systemd-unit-options.nix]: My fear was that without it, it would lead to module local hacks instead - this is how I personally solved it in a yet-to-be-finished module that I worked on before this. Maybe I'm wrong, though. I would of course prefer if these options weren't needed at all.

File: nixos/modules/services/misc/gitlab.nix
Comment (talyz) [reply nixos/modules/services/misc/gitlab.nix]: Yeah, I agree. I went with this approach since it seemed like the norm for NixOS modules (like gitea) and this was hinted at in the old PR.

File: nixos/modules/services/misc/gitlab.nix
Comment (aanderse) [reply nixos/modules/services/misc/gitlab.nix]: > Any password files should already have read permissions for the user that requires them.

My intention was to point out operations involving the `passwordFile` should not require `root` privileges, generally speaking.

> The problem is that there's no way, afaict, to create a user with a password in a secure fashion using the existing postgres service options. Maybe the postgres module should provide a way, though?

I'm suggesting that passwords are inferior to socket authentication, so socket authentication should be preferred when provisioning a database locally.

> Same here, as you say, it's not currently possible, but maybe it could be implemented?

From the looks of your code it looks like it could be implemented fairly easily. I believe I've seen extensions provisioned this way in several modules, so I think this would be valuable.

> Yeah, I guess we could introduce a createLocally option and rely on it instead of checking databaseHost.

Some people may wish to provision their own database, or be migrating from an existing install so `createLocally` is nice to have.

File: nixos/modules/system/boot/systemd-unit-options.nix
Comment (aanderse) [reply nixos/modules/system/boot/systemd-unit-options.nix]: I'm always curious to see situations where something can't be achieved in a declarative way with NixOS because then we can see patterns emerge and come up with declarative solutions, like the author of `services.mysql.ensureDatabases` did.

Keep in mind that module local hacks can always be cleaned up, but once you add `*FullPrivileges` functions you're stuck with that interface for better or worse.

File: nixos/modules/services/misc/gitlab.nix
Comment (aanderse) [reply nixos/modules/services/misc/gitlab.nix]: You're entirely correct it is the norm for NixOS modules and this needs to be fixed! I'm hoping to remove a bunch of these options before `19.09`.

File: nixos/modules/services/misc/gitlab.nix
Comment (talyz) [reply nixos/modules/services/misc/gitlab.nix]: Hm, yeah, I guess they could be generated on first run. This would however be less declarative and hide state that is essential to read any backups made of the gitlab instance, so while it's initially more convenient, it might be worse in the long run...

File: nixos/modules/services/misc/gitlab.nix
Comment (talyz) [reply nixos/modules/services/misc/gitlab.nix]: Yeah, sounds like a good idea. :)

File: nixos/modules/services/misc/gitlab.nix
Comment (aanderse) [reply nixos/modules/services/misc/gitlab.nix]: Modules like this are intrinsically stateful. There is a database. There is a `stateDir`. If a sysadmin fails to backup any of the state this application generates/utilizes the application becomes useless.

I would suggest these keys are implementation details not worth bothering the sysadmin over.

File: nixos/modules/services/misc/gitlab.nix
Comment (talyz) [reply nixos/modules/services/misc/gitlab.nix]: > My intention was to point out operations involving the `passwordFile` should not require `root` privileges, generally speaking.

Ah, I see.

> I'm suggesting that passwords are inferior to socket authentication, so socket authentication should be preferred when provisioning a database locally.

True, but socket authentication isn't always possible. For gitlab it seems like it is, though, so I should experiment a bit and see if a switch to it is doable.

> From the looks of your code it looks like it could be implemented fairly easily. I believe I've seen extensions provisioned this way in several modules, so I think this would be valuable.

Yes, probably. I'll have to look at the postgres module a bit and see what I can come up with.

Comment (aanderse) [reply nixos/modules/services/misc/gitlab.nix]: > True, but socket authentication isn't always possible.

And in those extremely rare cases we *might* consider having the sysadmin provision the database, or write some other manual code, consider it on a case by case basis because it is so rare.

> Yes, probably. I'll have to look at the postgres module a bit and see what I can come up with.

That's great. If you need a hand ping me.

File: nixos/modules/services/misc/gitlab.nix
Comment (talyz) [reply nixos/modules/services/misc/gitlab.nix]: What I was hinting at was that the backups done using the rake tasks would backup everything _but_ the secrets. I'm guessing that's why they weren't just automatically generated like the `gitlab_shell_secret` is. To always generate them also seems a bit contradictory to what you want with `createLocally`, or do you just want nothing to be created at all if `createLocally` is false?

File: nixos/modules/system/boot/systemd-unit-options.nix
Comment (talyz) [reply nixos/modules/system/boot/systemd-unit-options.nix]: That's true - I guess one way forward would be to convert this into a local hack and look at cleaning it up later, since I would prefer if this could be done in time for 19.09...

File: nixos/modules/services/misc/gitlab.nix
Comment (aanderse) [reply nixos/modules/services/misc/gitlab.nix]: I'll clarify I mean `services.gitea.database.createLocally` which is very separate than if you had a `services.gitea.createLocally` (I'm not sure what that might imply). A `database.createLocally` only implies that a local database and database user should be provisioned.

I'm not overly familiar with the `gitlab` module so if I'm wrong about automatically creating secrets please say so. Maybe the automated backups imply automatically generating secrets is a bad idea... though we didn't think that was the case with `gitea`.

File: nixos/modules/system/boot/systemd-unit-options.nix
Comment (aanderse) [reply nixos/modules/system/boot/systemd-unit-options.nix]: If we buckle down on this I think we could get it done in time for `19.09` in a declarative way, without the need for escalated scripts. I'm happy to chip in.

File: nixos/modules/services/misc/gitlab.nix
Comment (talyz) [reply nixos/modules/services/misc/gitlab.nix]: Okay, then I'll just remove the old options.

File: nixos/modules/services/misc/gitlab.nix
Comment (talyz) [reply nixos/modules/services/misc/gitlab.nix]: Yes, I got that, but at least two of the secrets are used to encrypt contents of the database. In a comment in another thread you said:

>Some people may wish to provision their own database, or be migrating from an existing install so `createLocally` is nice to have.

Which wouldn't be possible unless the secrets are also configurable.

Although I would agree that it's nice to just have to set `services.gitlab.enable = true` to get gitlab up and running, I feel like it would come at too high a cost in this case:

- Not offering the secrets as options makes migration and manual provision tricky and is potentially hostile to beginners who don't know of their importance
- Generating only unset secrets opens up a new can of worms of user errors (not that this was suggested, but I thought of it as an alternative)

What I can think of as a reasonable middle ground would be to introduce a new option `generateSecrets`, which would control this behavior, and have it turned off by default. Maybe this should also be done for gitea, erring on the side of control over magic, but I don't know enough about gitea to say. :)

File: nixos/modules/services/misc/gitlab.nix
Comment (lheckemann) [reply nixos/modules/services/misc/gitlab.nix]: @aanderse `after` doesn't mean it depends on it, see `systemd.unit(5)`

Comment (aanderse) [reply nixos/modules/services/misc/gitlab.nix]: @lheckemann I understand but had a poor choice of wording in this case. Sorry. When I said "depends" I didn't mean it in the `systemd` sense of the word, I just meant it was listed, when though not being used. Always nice to have clear and concise code.

File: nixos/modules/services/misc/gitlab.nix
Comment (aanderse) [reply nixos/modules/services/misc/gitlab.nix]: > it's nice to just have to set `services.gitlab.enable = true` to get gitlab up and running

That should generally be one of the top goals of all module authors. The beauty of NixOS lies right there :smile:

That being said... you raise some really good points. I'm still inclined to think that restoring backups might involve some work from the sysadmin (and every sysadmin is already taking backups of state regularly I sure hope) so I'm less concerned about the drawbacks than you are. As far as secrets options go I've said my piece and will defer to your judgement on the matter. Thanks for hearing me out :smile:

File: nixos/modules/services/misc/gitlab.nix
Comment (talyz) [reply nixos/modules/services/misc/gitlab.nix]: I managed to get gitlab up and running using peer authentication and pushed a quick draft using it. The extension stuff is still TBD and I'll probably rebase this to get rid of the old stuff, but still, if you could take a look and see if it aligns with your expectations, that would be great. :)

Comment (aanderse) [reply nixos/modules/services/misc/gitlab.nix]: Only suggestion I might have is to save some typing and possibly use `optionalAttrs` on the whole thing, like [this](https://github.com/NixOS/nixpkgs/blob/c0e56afddbcf6002e87a5ab0e8e17f381e3aa9bd/nixos/modules/services/monitoring/zabbix-server.nix#L216). But yeah, exactly what I was thinking. Good stuff :+1:

**Comment by flokli** [comment]
Thanks for all the great work done here :-)

Can you rebase test and rebase this on top of https://github.com/NixOS/nixpkgs/pull/66621?

**Comment by talyz** [comment]
@flokli Thanks! Sorry for the late reply - I've been on vacationing in Berlin for the last week, but I'll be going to cccamp now, so I'll probably have an opportunity to finish this up. I can see that the pr is merged, so I'll just rebase on master. In the meantime I'll push what I have at the moment and anyone interested can have a look. :)

**Comment by talyz** [comment]
Hm, just noticed that ´services.postgresql.ensureDatabases´ isn't in stable, so the backwards compatibility stuff I implemented is probably not necessary. I'll get rid of it.

File: nixos/modules/services/databases/postgresql.nix
Comment (aanderse) [inline nixos/modules/services/databases/postgresql.nix]: @talyz I'm not a fan of how `createDatabases` is *not* reproducible, where `ensureDatabases` was.

Comment (talyz) [reply nixos/modules/services/databases/postgresql.nix]: Hm, I'm not sure what you mean by that - in what way is it less reproducible than `ensureDatabases`?

Comment (aanderse) [reply nixos/modules/services/databases/postgresql.nix]: The name of the option `create` vs `ensure` already says half the story. If I modify the `owner` in `createDatabases` after I've already deployed once the database won't be created as described. Hence the rename to `create` makes sense. This is one reason why `ensureDatabases` is more appealing than `createDatabases`. Another reason being compatibility with `services.mysql.ensureDatabases`.

You'll have to forgive my lack of fluency in `postgresql`, so if what I'm saying is wrong please let me know why. What are the implications of the current implementation of `services.postgresql.ensureDatabases` as it stands if you were to use it in this module?

Comment (talyz) [reply nixos/modules/services/databases/postgresql.nix]: The name was actually chosen because I wanted to implement it in a backwards compatible way, but if you would prefer it, I can keep the `ensureDatabases` name, since I think the backwards compatibility should be dropped anyway. One reason to keep the `createDatabases` name would, as you hint at, be to differentiate it to `services.mysql.ensureDatabases`; I don't see why the functionality of the postgres module should be held back to be compatible with the mysql module.

What I could do is to implement owner switching - I didn't really have time to research it before, but it seems doable. That should get it up to the same level of reproducibility as `ensureDatabases`.

Comment (flokli) [reply nixos/modules/services/databases/postgresql.nix]: It's about more generalizing the database part in each nixos module, see the discussion starting at https://github.com/NixOS/nixpkgs/pull/56720#issuecomment-473255201.

Comment (talyz) [reply nixos/modules/services/databases/postgresql.nix]: I've implemented owner switching now, so it should be more reproducible.

I don't really know how to address the concern of keeping the interface similar to the mysql one, and at the moment there doesn't seem to be an answer to how the generalized interface would work. I can imagine the following solutions:

1. The interfaces are allowed to differ and the general interface wraps them as necessary
2. The interfaces have to be identical and we add a new `ensureExtensions` option
3. The interfaces have to be identical and we change the mysql module to match the postgresql one
4. We don't touch the postgresql module and come up with a different solution (the original approach, for example)
5. Something else...?

To me, solution 1 and 3 are obviously preferable. Solution 2 would be pretty ugly - extensions are added at database level, not globally. I think it would be a shame not to have `createDatabases`, since it provides an easy way to define database properties that can trivially be extended in the future.

File: nixos/modules/services/misc/gitlab.nix
Comment (aanderse) [inline nixos/modules/services/misc/gitlab.nix]: FYI https://github.com/NixOS/nixpkgs/pull/67394

Comment (talyz) [reply nixos/modules/services/misc/gitlab.nix]: Good to know. Thanks!

**Comment by talyz** [comment]
@flokli I rebased on master yesterday and it seems to work well. I've tried just switching to it, and resetting by deleting the db and state, and can't find any obvious issues. :)

File: nixos/modules/services/misc/gitlab.nix
Comment (aanderse) [inline nixos/modules/services/misc/gitlab.nix]: Is this needed with `createDatabaseLocally`? I thought socket authentication was used in that scenario.

**Review by aanderse — CHANGES_REQUESTED** [review]

File: nixos/modules/services/misc/gitlab.nix
Comment (talyz) [reply nixos/modules/services/misc/gitlab.nix]: Ah, no, that's a leftover from before createDatabaseLocally, I think. I've updated it to check for whether we're using socket authentication or not now.

File: nixos/modules/services/databases/postgresql.nix
Comment (talyz) [reply nixos/modules/services/databases/postgresql.nix]: This was resolved by moving the necessary stuff from the postgres module over to the gitlab module: we now inject what's needed into `systemd.services.postgresql.postStart` from there. The hope is that a better general database api will make this redundant by 20.03.

File: nixos/modules/services/misc/gitlab.nix
Comment (globin) [inline nixos/modules/services/misc/gitlab.nix]: Why don't you just remove the `default = null`? Then you don't need to create the assertions, same for the assertions below.

**Review by globin — CHANGES_REQUESTED** [review]

File: nixos/modules/services/misc/gitlab.nix
Comment (talyz) [reply nixos/modules/services/misc/gitlab.nix]: I did this because the generic message you get when you use an undefined parameter is less helpful, but if you want them gone, I can remove them.

**Review by aanderse — APPROVED** [review]
@talyz thanks for your work based on my comments. I'm approving the database parts of this change. I don't know much about `gitlab` on NixOS as I have never run the module (or looked at it much, before this PR) so I'll leave final approval to those more aware than myself.

Given the freeze coming up very quickly I would suggest @globin or @fpletz as the escalation point. @globin are you able to approve and merge this, or specify what changes are required to meet approval?

File: nixos/modules/services/misc/gitlab.nix
Comment (flokli) [inline nixos/modules/services/misc/gitlab.nix]: I think these two functions, plus the pattern to insert secrets are much more useful for nixos secret management in general, to be hidden in the gitlab module only. Thanks a lot for writing it :heart:

Could this be moved to `nixos/lib/utils.nix`, with maybe a line of docs in `nixos/doc`? I think other services can make use of that, too…

Comment (flokli) [reply nixos/modules/services/misc/gitlab.nix]: That being said, this should not block the PR before the feature freeze. If you don't find the time to do it in this PR, it can still be moved there at a later point.

Comment (talyz) [reply nixos/modules/services/misc/gitlab.nix]: No problem! :) I'll push the changes in a bit, but where in the docs would this fit?

Comment (talyz) [reply nixos/modules/services/misc/gitlab.nix]: @flokli I simplified the usage a bit, so that the pattern of using the functions together and outputting the JSON to a file for input is no longer necessary. Does it look okay?

**Review by etu — APPROVED** [review]

File: nixos/modules/services/misc/gitlab.nix
Comment (flokli) [reply nixos/modules/services/misc/gitlab.nix]: Sorry, somehow I missed the notification. I guess it's fine for now to have it as multi-line comments. Thanks!

Comment (talyz) [reply nixos/modules/services/misc/gitlab.nix]: No worries. :) Okay, great! Yeah, I think it should be in the docs, but I couldn't find a good spot. Maybe there should be a `Utility Functions` subsection to the `Functions reference` in the nixpkgs manual?

**Comment by srhb** [comment]
This broke master eval; `error: undefined variable 'genJqSecretsReplacementSnippet'' at /home/sarah/src/nixpkgs/nixos/lib/utils.nix:112:36`

**Comment by srhb** [comment]
Hmm, maybe it got lost in the merge?

**Comment by talyz** [comment]
@srhb Oops, yeah, that's my bad - I didn't rebase recently enough. Thanks for fixing it!

**Pull request merged**
